File: C/src/logger.c
#include "logger.h"
#include <stdio.h>
#include <stdarg.h>
#include <time.h>

static FILE* log_file = NULL;

void close_logger(void){
    if(log_file){
        fclose(log_file);
        log_file = NULL;
    }
}

void log_file_processing(const char* file_path, int success)
{
    LogLevel level = success ? LOG_INFO : LOG_ERROR;
    const char* status = success ? "SUCCESS" : "FAILED";
    log_message(level, "File Processing: %s - %s", file_path, status);
}


void init_logger(const char* filename) {
    log_file = fopen(filename, "w");
    if (!log_file) {
        fprintf("Failed to open the log file %s\n", filename);
    }
}

void log_message(LogLevel level, const char* format, ...){
    if (!log_file) return;

    const char* level_str;

    //Convert message level to human readable format
    switch (level) {
        case LOG_INFO: level_str = "INFO"; break; 
        case LOG_WARNING: level_str = "WARNING"; break;
        case LOG_ERROR: level_str = "ERROR"; break;
        case LOG_DEBUG: level_str = "DEBUG"; break;
        default: level_str = "UNKNOWN"; break;
    }

    //Convert time to human readable
    time_t now = time(NULL); 
    struct tm* tm = localtime(&now); 
    char time_str[20];
    strftime(time_str, sizeof(time_str), "%Y-%m-%d %H:%M:%S", tm);

    //Save log messages to the log file
    fprintf(log_file, "%s - %s - ", time_str, level_str);
    va_list args;
    va_start(args, format);
    vfprintf(log_file, format, args);
    va_end(args); 
    fprintf(log_file, "\n"); 
    fflush(log_file); 

    //Log messages to the console
    printf("%s - %s - ", time_str, level_str);
    va_start(args, format);
    vprintf(format, args);
    va_end(args);
    printf("\n");
}

File: logger.py
from datetime import datetime
from enum import Enum

class LogLevel(Enum):
    LOG_INFO = 1
    LOG_WARNING = 2
    LOG_ERROR = 3
    LOG_DEBUG = 4

log_file = None
LOG_FILE = "Log.log" #Log file name (and path!)

def log_summary(processed, failed):
    """Final log """
    log_message(LogLevel.LOG_INFO, f"Summary: Processed {processed}, Failed {failed}")

def close_logger():
    """Closes the log file"""
    global log_file
    if log_file:
        log_file.close()
        log_file = None

def log_file_processing(file_path, success):
    """
    Logs the file processing status — passed or failed.
    Parameters:
        file_path (str): Path to the file being processed.
        success (bool): True if compression succeeded
    """
    level = LogLevel.LOG_ERROR if not success else LogLevel.LOG_INFO
    status = "SUCCESS" if success else "FAILED"
    log_message(level, f"File Processing: {file_path} - {status}")

def init_logger(filename, mode = "w"):
    """
    Opens the log file and prepares the logger for use.
    Parameters:
        filename (str): Path to the log file.
    """
    global log_file
    try:
        log_file = open(filename, mode)
    except OSError as e:
        log_message(LogLevel.LOG_ERROR, f"Failed to open the log file {filename}: {e}")
        log_file = None

def log_message(level, format_str, *args):
    """
    Handles logging of formatted messages to both console and file.
    Parameters:
        level (LogLevel): Logging level (INFO, WARNING, ERROR, DEBUG).
        format_str (str): Format string for the log message.
        *args: Optional arguments to format into the message.
    """
    global log_file
    if not log_file:
        return
    
    level_map = {
    LogLevel.LOG_INFO: "INFO",
    LogLevel.LOG_WARNING: "WARNING",
    LogLevel.LOG_ERROR: "ERROR",
    LogLevel.LOG_DEBUG: "DEBUG" 
    }
    level_str = level_map.get(level, "UNKNOWN")

    time_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    message = format_str % args if args else format_str
    formatted = f"{time_str} - {level_str} - {message}"

    print(formatted)
    log_file.write(formatted + "\n")
    log_file.flush()